### DISCUSSION TITLE
How to use autopilot in UE5?

### DISCUSSION THREAD

**Discussion by cxr001** [JSBSim-Team/jsbsim#813] [Q&A]
I don't find a method to set roll pitch yaw command in c++, is that I need to write interface in FCS?  How to use the ap xml code in c++?

**Comment by seanmcleod** [comment]
Typically any autopilot implementation will export a number of properties, e.g.

https://github.com/JSBSim-Team/jsbsim/blob/5ac73c2b555ac7b6b97a1ef43f3f47f1f68a6f0f/aircraft/c172x/c172ap.xml#L27-L36

So you would then use the `FGPropertyManager` to write to for example the relevant setpoint, and to enable specific autopilot modes etc. For example here is some example C++ code using the PropertyManager, in this case to read the value of a property, in your case you would be writing to the autopilot specific properties.

```c++
FGPropertyManager* pm = fdm->GetPropertyManager();
FGPropertyNode* node = pm->GetNode("/my/property/name");

while (fdm->Run()) {
  x = node->getDoubleValue();
}
```

**Reply by cxr001** [reply]
So I can use Tie function in FGFCS::bind, for example, write 
PropertyManager->Tie("ap/altitude_setpoint", this, &FGFCS::GetAlCmd, &FGFCS::SetAlCmd); ?
But where is the attitude command here? I cannot find the pitch yaw command interface in c172ap, and it looks the flight controller doesn't use the airspeed setpoint to compute anything here.  Is that I should change this autoflight by myself?

**Comment by seanmcleod** [comment]
No, you would use the PropertyManager to write a value to `ap/altitude_setpoint`, e.g.

```c++
FGPropertyManager* pm = fdm->GetPropertyManager();
FGPropertyNode* node = pm->GetNode("ap/altitude_setpoint");

// Set the altitude set-point to 2000ft
node->setDoubleValue(2000);
```

**Reply by cxr001** [reply]
I can only find fdm from the FCS.cpp file, I don't know where is the fdm in the UE project

**Reply by cxr001** [reply]
I change the function copy to jsbsim as

void UJSBSimMovementComponent::CopyToJSBSim()
{
	// Basic flight controls
	FGPropertyManager* pm = Exec->GetPropertyManager().get();
	FGPropertyNode* node = pm->GetNode("ap/altitude_setpoint");
	// Set the altitude set-point to 2000ft
	node->setDoubleValue(2000);
	// FCS->SetDaCmd(Commands.Aileron);
	// FCS->SetRollTrimCmd(Commands.RollTrim);
	// FCS->SetDeCmd(Commands.Elevator);
	// FCS->SetPitchTrimCmd(Commands.PitchTrim);
	// FCS->SetDrCmd(-Commands.Rudder); // Rudder
	// FCS->SetDsCmd(Commands.Rudder); // Steering
	// FCS->SetYawTrimCmd(-Commands.YawTrim);
	// FCS->SetDfCmd(Commands.Flap);
	// FCS->SetDsbCmd(Commands.SpeedBrake);
	// FCS->SetDspCmd(Commands.Spoiler);
	  
	// Gears and Brake controls
	FCS->SetLBrake(FMath::Max(Commands.LeftBrake, Commands.ParkingBrake));
	FCS->SetRBrake(FMath::Max(Commands.RightBrake, Commands.ParkingBrake));
	FCS->SetCBrake(FMath::Max(Commands.CenterBrake, Commands.ParkingBrake));
	FCS->SetGearCmd(Commands.GearDown); 

	ApplyEnginesCommands(); 


But it doesn't work. Why?

**Comment by seanmcleod** [comment]
> But it doesn't work. Why?

What doesn't work? What are you expecting to happen, and what do you see happening?

**Reply by cxr001** [reply]
I want to control the airplane by autopilot, but it looks not change its attitude automatically. I think the autopilot was not enabled.

**Reply by cxr001** [reply]
As I give a setpoint for altitude, it should change its pitch value and try to fly higher I think

**Reply by cxr001** [reply]
I found that I have tie the "altitude_setpoint" in the JSBsimForUnreal project at first, then the code of  
node->setDoubleValue(2000); 
would not produce a flash back in UE, although it still doesn't change the UAV's attitude. If I use 
FGPropertyManager* pm = Exec->GetPropertyManager().get();
FGPropertyNode* node = pm->GetNode("ap/altitude_setpoint");
in  UJSBSimMovementComponent::CopyToJSBSim() directly, it will cause an address error. Maybe I write these codes in a wrong place, I am not fully understand.

**Reply by bcoconni** [reply]
> If I use FGPropertyManager* pm = Exec->GetPropertyManager().get(); FGPropertyNode* node = pm->GetNode("ap/altitude_setpoint"); in UJSBSimMovementComponent::CopyToJSBSim() directly, it will cause an address error. Maybe I write these codes in a wrong place, I am not fully understand.

Have you checked that `node` is not NULL before using it ? By default, `FGPropertyManager::GetNode` does not create a new node when the requested node does not already exist.

If `pm->GetNode()` returns `NULL` then it means that your aircraft has not defined the property `ap/altitude_setpoint`. I'd first check that the aircraft FDM (i.e. XML files) contain an autopilot as JSBSim does not supply a default autopilot: it's up to the FDM designer to create the autopilot for their aircraft.

**Reply by cxr001** [reply]
When I use 	
```c++
FGPropertyManager* pm =  FCS->GetPropertyManager().get(); //(or Exec->GetPropertyManager().get();)
FGPropertyNode* node_rh = pm->GetNode("ap/heading_setpoint");
```
the node_rh will return NULL. But I use the model of c172x, it should already include the autopilot. So I don't know how to repair my codes.

Sometimes node_al = pm->GetNode("ap/altitude_setpoint"); will return an address, but not controlling the elevator.

**Reply by gallonmate** [reply]
These functions already exist in the UE plugin and also work in Blueprints. I added them many months ago.
https://github.com/JSBSim-Team/jsbsim/blob/9f92a0603498ff09dd84e77b6e6022b866fcb50f/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L394
https://github.com/JSBSim-Team/jsbsim/blob/9f92a0603498ff09dd84e77b6e6022b866fcb50f/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L83-L115
Read Me File: https://github.com/JSBSim-Team/jsbsim/tree/9f92a0603498ff09dd84e77b6e6022b866fcb50f/UnrealEngine#extended-commands-and-properties
![image](https://user-images.githubusercontent.com/28702753/215239467-691261b7-b3ae-4e22-9ef2-d17d254c180b.png)

**Reply by cxr001** [reply]
I found that the roll channel can be controlled by ap/autopilot-roll-on, and also the ap/roll-attitude-mode. However, it looks, the attitude hold and heading hold still would not have any effect.

**Reply by gallonmate** [reply]
Yes I was curious so I tested the c172x and could not find any full working combination of settings for the auto pilot. But I tested c310 and seems that works good.

I came to the same conclusion as you with the c172x, only Altitude control works. The ap/autopilot-roll-on and also the ap/roll-attitude-mode when enabled do seem to stabilize the aircraft but it defaults to a heading of 103, I tried to override some properties to change the heading but I gave up after a few tries. The Auto Pilot system is confusing. I looked at the test scripts, some of which test the auto pilot of the c172x and found the functions do not work as suggested in the scripts. 
https://github.com/JSBSim-Team/jsbsim/blob/master/scripts/c1723.xml
https://github.com/JSBSim-Team/jsbsim/tree/master/scripts

In short I think c172x autopilot became broken in some update. The c310 seems to work good, but I only tested Heading and Altitude auto pilot controls.

Edit: For some reason, you also need to manually use engine starter in UE5 to start the engines on the c310. Press keys Control+Q or in Blueprint there is already Set Engines Starter in example BP_Airliner you can attach to Begin Play.

**Reply by cxr001** [reply]
As c310 is not suitable for my project, I checked the xml files again, and found that there is no "ap/aileron_cmd" property in c172r or c172x. Therefore the c172ap or other autopilot can not control the roll or heading positions of  c172.

**Comment by gallonmate** [comment]
> As c310 is not suitable for my project, I checked the xml files again, and found that there is no "ap/aileron_cmd" property in c172r or c172x. Therefore the c172ap or other autopilot can not control the roll or heading positions of c172.

@cxr001 good catch!
After searching I found issue already exist but xml file was never fixed https://github.com/JSBSim-Team/jsbsim/issues/180, 
Try editing the xml yourself by using this suggestion https://github.com/JSBSim-Team/jsbsim/issues/180#issuecomment-493667803 by @bcoconni 
I'm not sure why it was never fixed in the repo.

**Comment by seanmcleod** [comment]
Hmm, why are there two auto-pilot roll commands for roll control? Or was there some sort of renaming which was only done in one location and not the other?

https://github.com/JSBSim-Team/jsbsim/blob/98535e88748db3939910bd6267bcf13d297c60bf/aircraft/c172x/c172x.xml#L381-L390

Looking at the autopilot file and searching through it I couldn't find `ap/roll-cmd-norm-output`, so in addition to uncommenting `ap/aileron_cmd` we should remove `ap/roll-cmd-norm-output`.

**Reply by cxr001** [reply]
What I know here is, ap/aileron_cmd is included in autopilot, fcs/aileron-cmd-norm is used for manual fight in UE.  ap/roll-cmd-norm-output may not be used.

And do you know how to control and set "throttle-cmd-norm" in xml file? Although the c172ap.xml defines <property> ap/throttle-cmd-norm </property> as an interface at first, but it looks this property is not used in any other place, so it would not have any effect on the c172x model. I cannot find the property to adjust the throttle command/position in c172x.xml as well.

**Reply by gallonmate** [reply]
As I mentioned before, there are example script/test files. Here's how to control throttle. 
https://github.com/JSBSim-Team/jsbsim/blob/98535e88748db3939910bd6267bcf13d297c60bf/scripts/c1723.xml#L35-L38

**Reply by seanmcleod** [reply]
> I don't know whether UE5 can use scripts.

Take a look at the `InitializeJSBSim()` and `PrepareJSBSim()` methods in:

https://github.com/JSBSim-Team/jsbsim/blob/master/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp

I don't see any support for scripts currently.

**Reply by seanmcleod** [reply]
> This is a script file, 

Hmm, that doesn't look like a script file, take a look at the example script files in the scripts directory. The snippet you've shown comes from either the FDM file or a systems file.

**Reply by cxr001** [reply]
I think the code given by @gallonmate is a script, so I cannot use it. Forgive my poor English here. 
The code given by me couldn't work directly, I must read the output of the channel (the fcs/throttle-cmd-norm property), and then write the value into the dynamic by c++ or Blueprint. I don't know if this operation will cause any sampling problem. If there is any method or property to feedback the value of autopilot directly, I think it is better.

**Reply by gallonmate** [reply]
@cxr001 yes, sorry. I was confused and assumed you had a script working and just wanted to manually set throttle. There is not any default autopilot throttle. 

There is an autopilot throttle example in the Global5000ap.xml
But I have no idea if it works or how to use it.

**Reply by cxr001** [reply]
Thank you, I found it, named " fcs/throttle-pos-norm".  Note the throttle command value in UE would always display as 0 here.

**Comment by seanmcleod** [comment]
@cxr001 I've pushed a commit to fix the autopilot aileron command name in C172x - https://github.com/JSBSim-Team/jsbsim/commit/ae318dea96fcb1757e3d8b063cb6c21ef2d74dee
